src/stcal/ramp_fitting/utils.py

```diff
@@ -1317,28 +1319,63 @@ def dq_compress_final(dq_int, dnu_flag):
     # total number of integrations, that means at least one integration
     # has good data, so the final DQ flag for that pixel should NOT
     # include the DO_NOT_USE flag.
-    dnu = np.zeros(dq_int.shape, dtype=np.uint32)
-    dnu[np.where(np.bitwise_and(dq_int, dnu_flag))] = 1
-    dnu_sum = dnu.sum(axis=0)
-    not_dnu = np.where(dnu_sum < nints)
+    dnu_flag = ramp_data.flags_do_not_use
+    is_flag_set_for_all_ints(dq_int, f_dq, dnu_flag)
+
+    # Set the SATURATED flag in a similar way DO_NOT_USE is set.
+    sat_flag = ramp_data.flags_saturated
+    is_flag_set_for_all_ints(dq_int, f_dq, sat_flag)
 
-    not_dnu_flag = np.uint32(~dnu_flag)
-    f_dq[not_dnu] = np.bitwise_and(f_dq[not_dnu], not_dnu_flag)
+    # If the saturated flag is set, then the entire pixel is
+    # saturated, so should be marked as DO_NOT_USE.
+    sat_loc = np.zeros(f_dq.shape, dtype=f_dq.dtype)
+    sat_loc[np.where(np.bitwise_and(f_dq, sat_flag))] = 1
+    f_dq[sat_loc == 1] = np.bitwise_or(f_dq[sat_loc == 1], dnu_flag)
 
     return f_dq
 
 
+def is_flag_set_for_all_ints(dq_int, f_dq, flag):
+    """
```

Comment:
I had to stare at this function and its calls for a while to discern what it does. Maybe a bit more in the function description would help clarify it.

Reply:
I changed the docstring, as well as added more comments to clarify the purpose of this function.